zend_abstract_interface/symbols/lookup.c

```diff
@@ -0,0 +1,519 @@
+#include <sandbox/sandbox.h>
+#include "symbols.h"
+#include "zend_compile.h"
+
+// clang-format off
+static inline bool zai_symbol_update(zend_class_entry *ce ZAI_TSRMLS_DC) {
+    if (ce->ce_flags & ZEND_ACC_CONSTANTS_UPDATED) {
+        return true;
+    }
+
+    volatile bool zai_symbol_updated = true;
+
+    zai_sandbox sandbox;
+    zai_sandbox_open(&sandbox);
+
+    zend_try {
+#if PHP_VERSION_ID < 70000
+        zend_update_class_constants(ce ZAI_TSRMLS_CC);
+#else
+        if (zend_update_class_constants(ce) != SUCCESS) {
+            zai_symbol_updated = false;
+        }
+#endif
+    } zend_catch {
+        zai_symbol_updated = false;
+    } zend_end_try();
+
+    if (zai_symbol_updated && !EG(exception)) {
+        zai_sandbox_close(&sandbox);
+        return true;
+    }
+
+    zai_sandbox_close(&sandbox);
+    return false;
+}
+
+static inline void *zai_symbol_lookup_table(HashTable *table, const char *name, size_t length, bool ncase, bool pointer ZAI_TSRMLS_DC) {
```

Comment:
Once we drop PHP 5, we can make these things nicer with just the appropriate getter function like `zend_hash_str_find_ptr_lc` (polyfilling it for PHP 7)...